File: exchanges/kucoin.py
import time

import requests


def get_kucoin_futures_trade_link(currency1: str, currency2: str) -> str:
    return f"https://www.kucoin.com/ru/futures/trade/{currency1}{currency2}M"


def get_kucoin_funding_rates() -> dict[str, dict[str, float]]:
    funding_rates = requests.get(
        "https://api-futures.kucoin.com/api/v1/contracts/active"
    )

    return {
        funding_rate["symbol"][:-1]: {
            "funding_rate": funding_rate["fundingFeeRate"] * 100,
            "index_price": funding_rate["indexPrice"],
            "mark_price": funding_rate["markPrice"],
            "next_funding_time": time.time() + funding_rate["nextFundingRateTime"],
            "predicted_funding_rate": funding_rate["predictedFundingFeeRate"] * 100,
        }
        for funding_rate in funding_rates.json()["data"]
        if funding_rate["fundingFeeRate"]
    }

File: exchanges/kucoin.js
const axios = require('axios');

const { EXCHANGE_NAME } = require('../constants');
const { formatFundingRate, getFundingInterval, getTimeString } = require('../utils');

class KuCoin {
  getSpotTradeLink(symbol) {
    const currency = symbol.split('USDT')[0];
    return `https://www.kucoin.com/ru/trade/${currency}-USDT`;
  }

  getFuturesTradeLink(symbol) {
    const currency = symbol.split('USDT')[0];
    return `https://www.kucoin.com/ru/futures/trade/${currency}USDTM`;
  }

  async getFundingRates() {
    try {
      const { data: contracts } = await axios.get('https://api-futures.kucoin.com/api/v1/contracts/active');
      const fundingRates = {};

      for await (const contract of contracts.data) {
        const symbol = contract.symbol.slice(0, -1);

        if (typeof contract.fundingFeeRate === 'number') {
          try {
            const { data: fundingHistory } = await axios.get(
              `https://www.kucoin.com/_api_kumex/web-front/contract/${symbol}M/funding-rates?reverse=true&maxCount=1`
            );

            const nextFundingTime = Date.now() + contract.nextFundingRateTime;
            const fundingInterval = getFundingInterval(nextFundingTime, fundingHistory.data.dataList[0].timePoint);

            fundingRates[symbol.replace(/^10+/g, '')] = {
              fundingRate: formatFundingRate(contract.fundingFeeRate),
              indexPrice: contract.indexPrice,
              markPrice: contract.markPrice,
              nextFundingTime: getTimeString(nextFundingTime),
              fundingInterval,
              predictedFundingRate: formatFundingRate(contract.predictedFundingFeeRate),
              spotLink: this.getSpotTradeLink(symbol),
              futuresLink: this.getFuturesTradeLink(symbol),
              multiplier: symbol.match(/^10+/g)?.[0] ?? 1,
            };
          } catch (err) {
            console.log(`Ошибка обработки данных фандинга ${EXCHANGE_NAME.kucoin} (${symbol}). ${err?.message}`);
          }
        }
      }

      return fundingRates;
    } catch (err) {
      console.log(`Ошибка получения данных фандинга ${EXCHANGE_NAME.kucoin}. ${err?.message}`);
    }
  }
}

module.exports = new KuCoin();
